tests/federation_keys_test.go

```diff
@@ -29,60 +29,226 @@ import (
+					}
+
+					if v.Type != gjson.String {
+						return fmt.Errorf("signatures: Key '%s' for server '%s' has unexpected type, expected String, got '%s'", key, k.Str, v.Type.String())
+					}
+
+					_, err = base64.RawStdEncoding.DecodeString(v.Str)
+					if err != nil {
+						return fmt.Errorf("old_verify_keys: Key '%s' failed to decode as ed25519 base64: %s", k.Str, err)
+					}
 				}
-				// slightly evil to side-effect like this, but it's the easiest way
-				keyID = k.Str
-				key = keyBytes
+
 				return nil
 			}),
 		},
 	})
+
 	jsonObj := gjson.ParseBytes(body)
 	gotTime := time.Unix(0, jsonObj.Get("valid_until_ts").Int()*1000*1000) // ms -> ns
 	wantTime := time.Now()
 	if gotTime.Before(wantTime) {
-		t.Errorf("valid_until_ts : timestamp is in the past: %s < %s", gotTime, wantTime)
+		t.Errorf("valid_until_ts: timestamp is in the past: %s < %s", gotTime, wantTime)
+	}
+
+	checkKeysAndSignatures(t, body, jsonObj, keys, oldKeys)
+}
+
+func checkKeysAndSignatures(t *testing.T, body []byte, jsonObj gjson.Result, keys, oldKeys map[string]ed25519.PublicKey) {
+	var err error
+
+	if len(keys) == 0 {
+		t.Fatalf("verify_keys: missing any ed25519: key")
 	}
 
-	if keyID == "" {
-		t.Fatalf("missing ed25519: key")
+	if len(oldKeys) > 0 {
+		// Check if any old key exists in the new keys
+		for k := range oldKeys {
+			if _, ok := keys[k]; ok {
+				t.Fatalf("old_verify_keys: Key '%s' exists in both old_verify_keys and verify_keys", k)
+			}
+		}
 	}
-	sigBase64 := jsonObj.Get(fmt.Sprintf("signatures.hs1.%s", keyID))
-	if !sigBase64.Exists() {
-		t.Fatalf("missing signature for hs1.%s", keyID)
+
+	sigObj := jsonObj.Get("signatures").Map()
+
+	// Test signatures object sanity
+	if _, ok := sigObj["hs1"]; !ok {
+		t.Fatalf("signatures: Did not contain server key for 'hs1'")
 	}
-	sigBytes, err := base64.RawStdEncoding.DecodeString(sigBase64.Str)
-	if err != nil {
-		t.Fatalf("unable to decode base64 signature: %s", err.Error())
+
+	if len(sigObj) > 1 {
+		var extraKeys []string
+		for k := range sigObj {
+			if k == "hs1" {
+				continue
+			}
+			extraKeys = append(extraKeys, k)
+		}
+		t.Fatalf("signatures: Contained more than one server keys, extra keys: %s", strings.Join(extraKeys, ", "))
 	}
 
-	bodyWithoutSig, err := sjson.DeleteBytes(body, "signatures")
+	sigServerObj := sigObj["hs1"].Map()
+
+	type sigWithKey struct {
+		signature []byte
+		key       ed25519.PublicKey
+		old       bool
+	}
+
+	var signatures = map[string]sigWithKey{}
+
+	// Test signatures for all verify_keys, these *have* to exist.
+	for keyName, keyBytes := range keys {
+		sigBase64 := jsonObj.Get(fmt.Sprintf("signatures.hs1.%s", keyName))
+		if !sigBase64.Exists() {
+			t.Fatalf("signatures: missing signature for Key '%s'", keyName)
+		}
+
+		if sigBase64.Type != gjson.String {
+			t.Fatalf("signatures: Signature for Key '%s' has unexpected type, expected String, got '%s'", keyName, sigBase64.Type.String())
```

Comment:
You can use `must.GetJSONFieldStr(t, bodyBytes, fmt.Sprintf("signatures.hs1.%s", keyName))` to get string values from keys, which does checks for type/existence.